src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -3012,7 +3012,10 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
               else pt // allow type slack (pos/6221)
           }
 
-          unwrapWrapperTypes(ptNorm baseType FunctionSymbol) match {
+          unwrapWrapperTypes((ptNorm baseType FunctionSymbol) match {
+            case NoType | ErrorType => ptNorm // fallback to known type if no wrapped type
+            case t => t
+          }) match {
```

Comment:
So after this PR, we'd just use what ever the type arguments are in the expected type. I really don't think this is correct, for example, after your PR:

```
class RS {
  def map(t: T[String, Int]): RS = new RS
}

class T[X, Y]

object Test {
  implicit def c(f: Int => String) = new T[String, Int]
  
  def main(args: Array[String]): Unit = {
    val rs = new RS
    rs.map { x => x.length }
  }
}
```

This will infer the type `String => Int` for the function literal, and then fail to type check

```
Test.scala:12: error: type mismatch;
 found   : String => Int
 required: T[String,Int]
    rs.map { x => x.length }
               ^
```

There's really no connection between the expected type `T[String, Int]` and the function literal, and we should not just assume because `[String, Int]` are the type arguments these types should be the function's argument / result types.

Reply:
I deliberately switched the type parameters to demonstrate that using these type parameters to drive type inference of the function literal is completely arbitrary.

There could also be an expected type that doesn't have two type parameters, for example:

```scala
class RS {
  def map(t: BigInt): RS = new RS
}

object Test {
  implicit def c(f: Int => Int): BigInt = 0
  
  def main(args: Array[String]): Unit = {
    val rs = new RS
    rs.map { x => x }
  }
}
```

`BigInt` doesn't have any type arguments, so there's no expected type for type-checking the function literal, so we end up with

```
Test.scala:10: error: missing parameter type
    rs.map { x => x }
             ^
```

Maybe that makes my point more clear?

Reply:
Your example strips out the typing information necessary to make the inference.  BigInt doesn't carry the necessary typing information.  So the compiler correctly returns an error.  This happens even with my fix because the type inference isn't possible.  Contrast that with the reproduction case where ScopedValue[I, O] carries the argument (I) and return type (O) with it which is why the type inference engine is able to determine the type of the function.  This is how the type inference engine works right now.  All I did is fix an issue that happens after the type inference engine infers this information.

Reply:
I understand what your change is doing, and why it fixes your example. But I believe that it's an arbitrary solution. To give another example, if the expected type is `Foo[String, Long, Double]`, the type checker implementation would use `String => Double` as expected type for the function literal – that seems pretty arbitrary to me.

If the expected type is some subtype of `FunctionN[T1, T2, ..., Tn, R]`, the type checker uses `(T1, T2, ..., Tn) => R` as expected type for type checking the function literal. This was probably necessary to make function literal inference practical, and it makes sense to me because there is a connection between `FunctionN` and function literals (the default type of a function literal is `FunctionN`, though literals can also instantiate other SAM types). But even that can lead to unexpected inference, for example in current 2.13.8:

```scala
scala> def f(l: List[String]) = 0
def f(l: List[String]): Int

scala> f(x => "")
           ^
       error: type mismatch;
        found   : Int => String
        required: List[String]
```

The compiler picked `Int` as parameter type because `List[String] <: (Int => String)`. Replacing `List` by `Option`, you get `error: missing parameter type`.

Extending that mechanism to parameters of any type constructor (not just `FunctionN`) seems wrong to me.

But also, type inference is not specified :-) So there can be other opinions on that topic. @smarter, for example?

Reply:
As a rule of thumb, I'd be wary of any further entangling of type inference with implicit conversions, it's already a source of intractable issues (https://github.com/lampepfl/dotty/issues/8803), and one of the motivation for trying to get rid or at least weaken the power of implicit conversions (https://contributors.scala-lang.org/t/can-we-wean-scala-off-implicit-conversions/4388). I recommend making `ScopedValue` a SAM type instead of relying on an implicit conversion from a function type.

Reply:
I am not extending any behavior, it already exists.  I feel perhaps I am not the right audience for this topic.  I am just somebody who needed a bug fixed.

Reply:
@smarter That solution doesn't work for me as ScopedValue already has several methods.

Reply:
So we have the old thing that has a problem and the new thing that isn't ready.  And even though I did fix it and took a couple of weeks to do so, you won't merge my change?  Even though it is so small?  Even after jumping through all the CI hoops?

Reply:
@hunterpayne As open source maintainers, we are responsible to our community of users, and to our community of maintainers who will be helping us maintain all merged code for years or even decades to come. A responsibility not to merge anything that we cannot sincerely agree should be merged. The level of effort that was put it into it doesn't enter to it and _shouldn't_ enter into it. I know it's rough to put a bunch of effort into something and have it come to nothing (except for the learning involved, not just for you but for all of us, collectively). I believe that if this was your project and you were in our shoes, you'd make the same decision.

If you sincerely believe this PR should be merged, you need to make a better case than "but I put a lot of effort into it". That won't fly in this repo or in any well-maintained open source project.

Reply:
1) People like me making PRs like these is maintaining software (and the easier you make it on them to commit the more PRs you will get)
2) The change only effects unwrapped types
3) More complex versions of the same reproduction case work with more complex types.  Why should an map[Option[String]] work and map[String] not work.  This improves the consistency of the inference causing more types to be inferred in the same way rather than treating primitive (among other types) differently
4) Having the same line of code throw multiple compiler errors for the same error isn't good. (From the changes to the unit test check files)

Reply:
I'm really sorry if this is a frustrating experience for you, especially in your first contribution :-/

In a type system, making an example work is not a good reason to merge a change. Type systems are built on a theory, and sometimes they stand in the way of usability - but breaking consistency always makes problems pop up in other places. I understand that this might not be a very convincing argument, especially given that inference doesn't have a specification in Scala. But even if the spec is not written down, there are principles in there, in the code and in the experience of the maintainers (@adriaan who wrote that code, @smarter whom I pinged and myself all did a PhD in the field of type systems and have been maintaining Scala compilers for years).

So I think we can't accept this change - but I'm open to discuss technical arguments explaining why the change is correct and principled, it's entirely possible that I'm missing something.

Reply:
This will be my last communications.  I am will never contribute to this project ever again.  I also will stop using Scala both in my professional and opensource work.  Good bye.